Show previous results in test results page

okurz opened the pull request:
Querying the database for former test runs of the same scenario is a rather
costly operation which we do not want to do for multiple test results at once
but only for each individual test result (1:1 relation).

Related issue: https://progress.opensuse.org/issues/10212

Screenshot of feature:
![screenshot_20160210_142024](https://cloud.githubusercontent.com/assets/1693432/12948308/7e915a3c-d001-11e5-840b-2f070c3cb8a5.png)

* test results: Add failure details to previous jobs
* Also use direct reference to db in "sub show"
* test results: Add tab with "previous results" in same scenario
* templates: test: Reload selected tab also when no modlist

Tested against production database dump from openqa.opensuse.org, querying with
```
time sudo -u geekotest OPENQA_CONFIG=local/stage OPENQA_DATABASE=stage script/openqa get '/tests/117629' >/dev/null 2>&1 | grep real
```
yields the following time increase before and after change
* *origin/master*: 0.8s
* *feature/label_and_previous_results*: 1.1s

coolo commented on lib/OpenQA/Scheduler/Scheduler.pm:
```diff
@@ -293,6 +293,33 @@ sub _job_get($) {
     return $job->to_hash(assets => 1);
 }
 
+sub job_settings_subquery {
```
I guess the memo was sent before you joined - but we declared Scheduler.pm as dead. Code should be removed from it, not added. So please put this code into the model.

coolo commented on lib/OpenQA/WebAPI/Controller/Test.pm:
```diff
@@ -284,14 +288,38 @@ sub show {
+    my @conds;
+    push(@conds, {'me.state' => 'done'});
+    push(@conds, {'me.result' => {-not_in => [OpenQA::Schema::Result::Jobs::INCOMPLETE_RESULTS]}});
+    # instead of searching for "previous jobs" it might be better to search for same or previous builds
```
"previous build" is unfortunately in reality very unreliable, so I think you don't really have an alternative to what you do

coolo commented on lib/OpenQA/WebAPI/Controller/Test.pm:
```diff
@@ -233,7 +233,7 @@ sub read_test_modules {
 
 sub show {
-    my ($self) = @_;
+    my ($self, $limit_previous) = @_;
```
there are no arguments to mojo controllers. So if you want to define an undefined variable, just do so :)

coolo commented:
please also rebase, so the diff is a bit smaller

coolo commented on templates/test/tr_job_result_details.html.ep:
```diff
@@ -0,0 +1,7 @@
+% if ($comments and $comments > 0) {
```
you're cheating :)

okurz replied:
I am sorry, I don't want to. I commented on this in the corresponding commit message:
https://github.com/okurz/openQA/commit/c0e258d1e16920393e51a6b091600584bc6e0030
> The 'if ($comment and $comments > 0)' is used to prevent warnings about
undefined if comments are not available at all but also not show notifications
if it is always defined by 0 as used in 'previous.html.ep'.

Don't know a better way except for needing to count the available comments within Jobs.pm#show like in sub overview.

aaannz replied:
Ah, there are `$comment` and `$comments`. Just spent 5min staring at this :) . What about `if ($comment and defined $comments)`?

coolo replied:
the cmoment talks about $comment , the code does not - it's $comments only

aaannz replied:
Ah, right. So what is the issue with plain `if ($comments)` again? Am I missing something?

okurz replied:
this code section is included in two locations, templates/test/previous.html.ep with the _resultset_ including comments as parameter (always defined but can be empty) and templates/test/tr_job_result.html.ep with _$res->{comments}_ as parameter which is a scalar (undef if no comments).
If I just check with `if ($comments)` then it will always be true for "previous.html" and always display the symbol. If I just check with `if ($comments > 0)` then it will yield a warning for "tr_job_result.html".

So what approach do you prefer?

coolo replied:
previous sets $comments to the number of comments, right? if ($comments) will be false if that number is 0. So where is the problem?

coolo replied:
[ that's the difference between if ($comments) and if (defined $comments) ]

aaannz replied:
As @coolo said. [L26](https://github.com/os-autoinst/openQA/pull/538/files#diff-2e3f373e0c4e8c21b3a158b6c5ed691cR26) clearly set scalar as `$comments`.

coolo commented on lib/OpenQA/WebAPI/Controller/Test.pm:
```diff
@@ -284,14 +288,38 @@ sub show {
+    push(@conds, {'me.id' => {-in => $subquery->get_column('job_id')->as_query}});
+
+    my %attrs = (
+        rows => $limit_previous //= 10,    # arbitrary limit of previous results to show
```
please move that assignment in a line of its own. this is very ugly IMO

coolo commented on t/ui/16-tests_previous_results.t:
```diff
@@ -0,0 +1,42 @@
+my $previous_results_header = $t->tx->res->dom->at('#previous #scenario')->all_text;
+is($previous_results_header, q/Results for opensuse-13.1-DVD-i586-textmode, limited to 10/, 'header for previous results with scenario');
+$get->element_exists('#res_99945',                'result from previous job');
+$get->element_exists('#res_99945 .result_passed', 'previous job was passed');
```
@sysrich completely offtopic: what's the signficance of was here? 'it passed' is past, 'it has passed' is perfect. What is 'was passed'? It's a lot in the internet, so it must be correct - but what is it? :)

okurz replied:
You can read this as either "past perfect" to describe an action that completed before another event after that which is also already in the past **or** you can read it as "the job was completed with the result of 'passed'" which is what I intended here.

coolo replied:
but past perfect is built with had. So it's either "the job had passed" as past perfect or "the job was passed" as past of be with the adjective 'passed'. So much about the grammar. But neither past perfect makes sense here as there is no other event - nor does past, as the job's state is passed even in the future. It won't change anymore. Anyway, I was just curious.

coolo commented:
BTW: I would be curious how this behaves on a production copy data

okurz commented:
> BTW: I would be curious how this behaves on a production copy data

yes, plan for today

okurz commented:
updated:

 * performance test with production database dump (see description)
 * rebased (now also showing labels on previous results)
 * javascript adjustment on results page: Reload selected tab also when no modlist

updated screenshot:
![openqa_previous_results_with_labels](https://cloud.githubusercontent.com/assets/1693432/13084881/ca8189ec-d4db-11e5-93a7-411ef4280da6.png)

@coolo, @sysrich *ping*

Soulofdestiny commented:
If a test with the same name ran on different machines (e.g. "allpatterns" and "allpatterns@zkvm", it founds multiple previous jobs
intentional or bug?

![screenshot from 2016-03-01 15-43-36](https://cloud.githubusercontent.com/assets/3983634/13430220/6b4cc678-dfc4-11e5-98a7-972478be114f.png)

sysrich commented:
I think it's a bug, but it's an awesome one ;)

@okurz please change the filter so the previous jobs filter based on machine used by the job, not arch :)

okurz commented:
tracked in https://progress.opensuse.org/issues/11034
